### DISCUSSION TITLE
Ergonomics of new Keypair

### DISCUSSION THREAD

**Discussion by AgeManning** [libp2p/rust-libp2p#3983] [Q&A] [answered]
I'm updating some dependencies. 

I have a situation, where I have a libp2p keypair and depending on its type need to convert it to a new type modifying the encoding. 

Essentially I used to match on the `Keypair` variant, then perform misc encoding/decoding depending on the type. 

It seems we now need to use the `try_into_*` functions. For example, check if its secp256k1 run the `try_into_secp256k1()` function, if that fails try the one in the result. See here: https://github.com/libp2p/rust-libp2p/blob/master/identity/src/keypair.rs#L115. Note that the error doesn't return the original type, so on error the original Keypair is now lost.

Because these functions consume `self`. If I want to check what kind of type the keypair, is I have to consume self, to get an error that tells me what the key is. This means I now have to clone my private keys (which I'd prefer not to do, keep the memory footprint low for private keys), in order to determine which kind of key I have and perform required logic. 

Some questions:
1. Am i missing something here?
2. If I'm not, maybe we could add a function that tells us the variant
3. Maybe have a function that converts without consuming self

**Comment by mxinden** [comment]
@thomaseizinger can you comment here?

**Reply by thomaseizinger** [reply, marked as answer]
Sure, we have an open issue about this: https://github.com/libp2p/rust-libp2p/issues/3649

@divagant-martian brought this to my attention already! We have a `KeyType` enum now, it just needs to be used for the `KeyPair` as well.

I haven't gotten around to it yet unfortunately. PRs welcome! Other ideas too but I think this one is a decent trade-off :)

**Comment by AgeManning** [comment]
oh woops, lol. We are probably upgrading the same code too. Ahaha. Sorry for the noise.

**Reply by thomaseizinger** [reply]
No worries! :)
